extensions/OpenAI/OpenAI/OpenAITextGenerator.cs

```diff
@@ -159,17 +165,33 @@ public async IAsyncEnumerable<string> GenerateTextAsync(
 
         await foreach (StreamingTextContent x in result.WithCancellation(cancellationToken))
         {
-            // TODO: try catch
-            // if (x.Metadata?["Usage"] is not null)
-            // {
-            //     var usage = x.Metadata["Usage"] as ChatTokenUsage;
-            //     this._log.LogTrace("Usage report: input tokens {0}, output tokens {1}, output reasoning tokens {2}",
-            //         usage?.InputTokenCount, usage?.OutputTokenCount, usage?.OutputTokenDetails.ReasoningTokenCount);
-            // }
-
-            if (x.Text == null) { continue; }
-
-            yield return x.Text;
+            TokenUsage? tokenUsage = null;
+
+            // The last message in the chunk has the usage metadata.
+            // https://platform.openai.com/docs/api-reference/chat/create#chat-create-stream_options
+            if (x.Metadata?["Usage"] is ChatTokenUsage { } usage)
+            {
+                this._log.LogTrace("Usage report: input tokens {0}, output tokens {1}, output reasoning tokens {2}",
+                    usage.InputTokenCount, usage.OutputTokenCount, usage.OutputTokenDetails?.ReasoningTokenCount ?? 0);
+
+                tokenUsage = new TokenUsage
+                {
+                    Timestamp = DateTime.UtcNow,
+                    ServiceType = "OpenAI",
+                    ModelType = "TextGeneration",
+                    ModelName = this._textModel,
+                    ServiceTokensIn = usage!.InputTokenCount,
+                    ServiceTokensOut = usage.OutputTokenCount,
+                    ServiceReasoningTokens = usage.OutputTokenDetails?.ReasoningTokenCount
+                };
+            }
+
+            // NOTE: as stated at https://platform.openai.com/docs/api-reference/chat/streaming#chat/streaming-choices,
+            // The Choice can also be empty for the last chunk if we set stream_options: { "include_usage": true} to get token counts, so it is possible that
+            // x.Text is null, but tokenUsage is not (token usage statistics for the entire request are included in the last chunk).
+            if (x.Text is null && tokenUsage is null) { continue; }
+
+            yield return (x.Text, tokenUsage);
```

Comment:
if text is null I'd return an empty string, so the caller doesn't have to worry about nulls